--- chrpath-0.9/chrpath.c~	Tue Sep 17 12:59:42 2002
+++ chrpath-0.9/chrpath.c	Tue Nov  5 16:00:48 2002
@@ -198,8 +198,6 @@
 
   printf("%s: %s=%s\n", filename, elf_tagname(dyns[rpath_dyns_index].d_tag),
          rpath);
-  free(dyns);
-  dyns = NULL;
 
   if (NULL == newpath)
   {
@@ -246,6 +244,8 @@
   }
   printf("%s: new %s: %s\n", filename,
          elf_tagname(dyns[rpath_dyns_index].d_tag), rpath);
+  free(dyns);
+  dyns = NULL;
 
   elf_close(fd);
 
